src/shims/fs.rs

```diff
@@ -722,6 +799,248 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+            builder.mode(_mode.into());
+        }
+
+        let result = builder.create(path).map(|_| 0i32);
+
+        this.try_unwrap_io_result(result)
+    }
+
+    fn rmdir(
+        &mut self,
+        path_op: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        this.check_no_isolation("rmdir")?;
+
+        let path = this.read_os_str_from_c_str(this.read_scalar(path_op)?.not_undef()?)?;
+
+        let result = remove_dir(path).map(|_| 0i32);
+
+        this.try_unwrap_io_result(result)
+    }
+
+    fn opendir(&mut self, name_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, Scalar<Tag>> {
+        let this = self.eval_context_mut();
+
+        this.check_no_isolation("opendir")?;
+
+        let name = this.read_os_str_from_c_str(this.read_scalar(name_op)?.not_undef()?)?;
+
+        let result = read_dir(name);
+
+        match result {
+            Ok(dir_iter) => {
+                let id = this.machine.dir_handler.insert_new(dir_iter);
+
+                // The libc API for opendir says that this method returns a pointer to an opaque
+                // structure, but we are returning an ID number. Thus, pass it as a scalar of
+                // pointer width.
+                Ok(Scalar::from_int(id, this.pointer_size()))
+            }
+            Err(e) => {
+                this.set_last_error_from_io_error(e)?;
+                Ok(Scalar::from_int(0, this.memory.pointer_size()))
+            }
+        }
+    }
+
+    fn readdir64_r(
+        &mut self,
+        dirp_op: OpTy<'tcx, Tag>,
+        entry_op: OpTy<'tcx, Tag>,
+        result_op: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        this.check_no_isolation("readdir64_r")?;
+
+        let dirp = this.read_scalar(dirp_op)?.to_machine_usize(this)?;
+
+        let entry_ptr = this.force_ptr(this.read_scalar(entry_op)?.not_undef()?)?;
+        let dirent64_layout = this.libc_ty_layout("dirent64")?;
+
+        let dir_iter = this.machine.dir_handler.streams.get_mut(&dirp).ok_or_else(|| {
+            err_unsup_format!("The DIR pointer passed to readdir64_r did not come from opendir")
+        })?;
+        match dir_iter.next() {
+            Some(Ok(dir_entry)) => {
+                // Write into entry, write pointer to result, return 0 on success.
+                // The name is written with write_os_str_to_c_str, while the rest of the
+                // dirent64 struct is written using write_packed_immediates.
+
+                let name_offset = dirent64_layout.details.fields.offset(4);
+                let name_ptr = entry_ptr.offset(name_offset, this)?;
+
+                let file_name = dir_entry.file_name();
+                let name_fits = this.write_os_str_to_c_str(&file_name, Scalar::Ptr(name_ptr), 256)?;
+                if !name_fits {
+                    panic!("A directory entry had a name too large to fit in libc::dirent64");
```

Comment:
I don't think `panic!` is a good reaction here -- not if it is possible for user code to actually reach this. It should be impossible to trigger such ICEs.
`throw_unsup_format!` seems more appropriate.